tokio/src/sync/semaphore_ll.rs

```diff
@@ -590,22 +748,37 @@ impl Permit {
     ///
     /// Repeatedly calling `forget` without associated calls to `add_permit`
     /// will result in the semaphore losing all permits.
-    pub(crate) fn forget(&mut self) {
-        self.forget2();
-    }
+    ///
+    /// Will forget **at most** the number of acquired permits. This number is
+    /// returned.
+    pub(crate) fn forget(&mut self, n: u16) -> u16 {
+        use PermitState::*;
 
-    /// Returns `true` if the permit was acquired
-    fn forget2(&mut self) -> bool {
         match self.state {
-            PermitState::Idle => false,
-            PermitState::Waiting => {
-                let ret = self.waiter.as_ref().unwrap().cancel_interest();
-                self.state = PermitState::Idle;
-                ret
+            Waiting(requested) => {
+                let n = cmp::min(n, requested);
+
+                // Decrement
+                let acquired = self
+                    .waiter
+                    .as_ref()
+                    .unwrap()
+                    .try_dec_permits_to_acquire(n as usize) as u16;
+
+                if n == requested {
+                    self.state = Acquired(0);
+                } else if acquired == requested - n {
```

Comment:
It does not look like this branch would be required

Reply:
Unfortunately it is as `n` can vary between calls (IMO this is a problem w/ the semaphore's API, but I didn't want to do a bigger refactor).